MessageSet tests + improvements

markrwilliams opened the pull request:
https://twistedmatrix.com/trac/ticket/9177

(This branch originally was also going to include https://twistedmatrix.com/trac/ticket/9178 and https://twistedmatrix.com/trac/ticket/9179, but it's larger than I anticipated so I'm putting it into review now.)

markrwilliams committed:
Move MessageSet tests into their own test case.

IMAP4HelperTests.test_messageSet becomes
MessageSetTests.test_equalityAndAddition, becauses that's what it
does.

markrwilliams committed:
100% coverage for MessageSet.

Dead code in _iterator, a private API, was eliminated; __iter__ checks
that the first element in the first range is None, and raises a
TypeError if it is.  None is smaller than all other objects, and
MessageSet.ranges is continually sorted via MessageSet.clean, so one
or more ranges that begin with None would cause the first range to
begin with None.  Consequently l in _iterator could never be None.

Implement __ne__ so that the isinstance check in __eq__ can fail with
!=, not just ==.

markrwilliams committed:
Topfile placeholder to make Travis pass.

markrwilliams commented on src/twisted/mail/test/test_imap.py:
```diff
@@ -296,6 +296,86 @@ def cbProduced(result):
 
 
+class MessageSetTests(unittest.SynchronousTestCase):
+    """
```
[This commit](https://codecov.io/gh/twisted/twisted/src/9f61890d97c41065dbb5cece2bd4af986c027363/src/twisted/mail/imap4.py) decreased [the previous commit's coverage](https://codecov.io/gh/twisted/twisted/src/562496b261a71b3372c7dd8015f556d32f1285cb/src/twisted/mail/imap4.py); [Two](https://codecov.io/gh/twisted/twisted/src/9f61890d97c41065dbb5cece2bd4af986c027363/src/twisted/mail/imap4.py#L279) [lines](https://codecov.io/gh/twisted/twisted/src/9f61890d97c41065dbb5cece2bd4af986c027363/src/twisted/mail/imap4.py#L282) aren't covered because this commit no longer passes the output of [`parsedIdList` to `MessageSet`](https://github.com/twisted/twisted/pull/817/commits/9f61890d97c41065dbb5cece2bd4af986c027363?diff=unified#diff-904151008804a75ba425f52b818531c8L399), which [doesn't make sense](https://twistedmatrix.com/trac/ticket/9178).

markrwilliams committed:
twisted.mail.imap4.MessageSet sorts * larger than any message.

RFC 3501 declares that * is not necessarily the largest UID, just the
most recent.  However treating * as the largest UID is consistent with
Cyrus' interpretation of *, and means None can be replaced with
float('inf') for all comparisons.

None now ends up in a consistent location in MessageSet.ranges, and is
never compared to an integer.  This makes the implementation more
robust and easier to port to Python 3, where numeric comparisons to
None fail with a TypeError.

glyph commented on src/twisted/mail/imap4.py:
```diff
@@ -209,33 +209,70 @@ def clean(self):
+                                max(previousHigh, high))
 
-        self.ranges = [r for r in self.ranges if r]
+        self.ranges = [(None if low is self._infinity else low,
```
I can't help but notice that this fairly complex comprehension occurs twice.  Put it into a function?

glyph commented on src/twisted/mail/imap4.py:
```diff
@@ -209,33 +209,70 @@ def clean(self):
+        @return: L{True} if L{None} is in some range in ranges and
+            L{False} if otherwise.
+        """
+        # The cleanup method merges overlapping or consecutive ranges.
```
Please put this lovely comment into the docstring.

glyph replied:
Although: what is "the cleanup" method? `L{MessageSet.cleanup}`?

glyph commented on src/twisted/mail/imap4.py:
```diff
@@ -209,33 +209,70 @@ def clean(self):
+        return self.ranges[-1][-1] is None
 
 
     def __contains__(self, value):
```
While you're in here, an `@type` on `value` would be very helpful.

glyph commented:
If I were reviewing this I'd probably give it a passing review.  Big jump in test coverage, comprehensible design, etc.  What do you think it still needs?

markrwilliams committed:
Tests to confirm the spec allows * to be the largest ID.

markrwilliams committed:
Documentation improvements (thanks, @glyph!)

markrwilliams committed:
Generalize swapping logic.

markrwilliams committed:
Remove unresolvable intra-docstring references.

markrwilliams committed:
Merge branch 'trunk' into 9177-9178-9179-messageset

glyph commented on src/twisted/mail/imap4.py:
```diff
@@ -127,27 +172,42 @@ def _setLast(self, value):
         return _getLast, _setLast, None, doc
     last = property(*last())
 
+    def _swap(self, this, that, ifIs):
```
This doesn't use `self` at all and should therefore be a top-level function, not a method.

glyph commented on src/twisted/mail/imap4.py:
```diff
@@ -209,33 +275,76 @@ def clean(self):
         """
 
-        self.ranges.sort()
+        ranges = sorted((self._swap(low, self._infinity, ifIs=None),
```
This same, fairly complex double-swap expression happens both at the top and the bottom of this function. It should be pulled out to a separate method (although I gather the `sorted` and `list` invocations can stay here.)

glyph commented on src/twisted/mail/imap4.py:
```diff
@@ -209,33 +275,76 @@ def clean(self):
-        oldl, oldh = None, None
-        for i,(l, h) in enumerate(self.ranges):
-            if l is None:
+        mergedRanges = [(float('-inf'), float('-inf'))]
```
perhaps this should be bound to a global called `_inf`?

(I also vaguely seem to remember that `1e999` is more portable than `inf` as a constant, but none of our buildbots are complaining, so…)

markrwilliams replied:
It looks like `inf`, `-inf`, and `nan` became portable [a while ago](https://bugs.python.org/issue1635):

> The float() function will now turn the string nan into an IEEE 754 Not A Number value, and +inf and -inf into positive or negative infinity. **This works on any platform with IEEE 754 semantics**. (Contributed by Christian Heimes; issue 1635.)

(from https://docs.python.org/2/whatsnew/2.6.html, emphasis added.)

So I think the infinities are safe...

markrwilliams replied:
I left these as literals because they're used only in this function as the base case for the loop that follows, whereas `MessageSet._infinity`'s unique identity allows `_swap` to reliably identify it.

glyph reviewed (approved):
Overall this looks quite good.  The buildbot error on the docs builder is unfortunately real, https://buildbot.twistedmatrix.com/builders/documentation/builds/2972/steps/api-documentation/logs/new%20pydoctor%20errors but the appveyor failure is spurious (since scrolling takes forever there, traceback recorded at https://twistedmatrix.com/trac/ticket/8879#comment:8).

markrwilliams committed:
MessageSet.cleanup doesn't exist, but .cleanup does.

markrwilliams committed:
_swap shouldn't be an instance variable, and swapping pairs can be generalized.

Per https://github.com/twisted/twisted/pull/817#discussion_r122338258
and https://github.com/twisted/twisted/pull/817#discussion_r122338566

markrwilliams committed:
Merge branch 'trunk' into 9177-9178-9179-messageset

markrwilliams committed:
tuples don't exist but tuple does
